### PR TITLE
add nettools sal

### PR DESCRIPTION
**Pull request by alichaddad**

### CODE REVIEW

File: jumpscale/sals/nettools/__init__.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
+# TODO: test *2
+
+import time
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: please let's use j.data.time, it's has a very friend time api using arrow library

File: jumpscale/sals/nettools/__init__.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+import socket
+import netaddr
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: in python3 there's ipaddress library that covers use cases in this library

File: jumpscale/sals/nettools/__init__.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+import re
+import os
+import urllib
+from jumpscale.god import j
+
+
+def tcp_connection_test(ipaddr: str, port: int, timeout=None):
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: you can say timeout:Optional[int] , make sure to import it from `typing`

File: jumpscale/sals/nettools/__init__.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    conn = socket.socket(type=socket.SOCK_DGRAM)
+    if timeout:
+        conn.settimeout(timeout)
+    try:
+        conn.connect((ipaddr, port))
+    except BaseException:
+        conn.close()
+        return False
+
+    print(f"Sending {len(message)} bytes to {ipaddr}:{port}")
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: j.logger instead of print

File: jumpscale/sals/nettools/__init__.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        timeout (int, optional): timeout of test. Defaults to 5.
+
+    Raises:
+        j.exceptions.Input: raises if not correct url
+
+    Returns:
+        bool: True if the test succeeds, False otherwise
+    """
+    try:
+        code = urllib.request.urlopen(url, timeout=timeout).getcode()
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: let's move to j.tools.http

File: jumpscale/sals/nettools/__init__.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+        return code == 200
+    except ValueError:
+        raise j.exceptions.Input("Please specify correct url with correct scheme")
+    except urllib.error.URLError:
+        return False
+
+
+def get_nics():
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: get_nic_names

File: jumpscale/sals/nettools/__init__.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+            if match and match.group("driver") == "bridge":
+                return "VLAN"
+            return "ETHERNET_GB"
+
+    elif j.data.platform.is_osx():
+        command = f"ifconfig {interface}"
+        exitcode, output, err = j.core.executors.run_local(command, hide=True, warn=True)
+        if exitcode != 0:
+            # temporary plumb the interface to lookup its mac
+            print(
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: j.logger & f-strings

File: jumpscale/sals/nettools/__init__.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+            if len(match.group("interfaceid")) >= 4:
+                return "VLAN"
+            else:
+                if len(interfacepieces) > 1:
+                    return "VIRTUAL"
+                else:
+                    return "ETHERNET_GB"
+
+
+def get_reachable_ip_address(ip: str, port: int):
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: you can add `-> str` in the function signature to indicate the return type of the function

File: jumpscale/sals/nettools/__init__.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+def ping_machine(ip, pingtimeout=60, allowhostname=True):
+    """Ping a machine to check if it's up/running and accessible
+    @param ip: Machine Ip Address
+    @param pingtimeout: time in sec after which ip will be declared as unreachable
+    @param allowhostname: allow pinging on hostname (default is false)
+    @rtype: True if machine is pingable, False otherwise
+    """
+    if not allowhostname:
+        try:
+            netaddr.IPAddress(ip)
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: ipaddress instead of netaddr

File: jumpscale/sals/nettools/__init__.py [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+        except netaddr.AddrFormatError:
+            raise j.exceptions.Value("Invalid ip address, set allowedhostname to use hostnames")
+
+    start = time.time()
+    pingsucceeded = False
+    while time.time() - start < pingtimeout:
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: j.data.time

File: jumpscale/sals/nettools/__init__.py (hunk 1 above)
Comment (alichaddad) [reply jumpscale/sals/nettools/__init__.py]: it doesn't have sleep function

File: jumpscale/sals/nettools/__init__.py [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+            exitcode, _, _ = j.core.executors.run_local(f"ping -c 1 {ip}", warn=True, hide=True)
+        if exitcode == 0:
+            pingsucceeded = True
+            return True
+        time.sleep(1)
+    if not pingsucceeded:
+        return False
+
+
+def download(url, localpath, username=None, passwd=None, overwrite=True):
```
Comment (xmonader) [inline jumpscale/sals/nettools/__init__.py]: should be upgraded to requests stream or something better

**Pull request merged**